### PR TITLE
Make plugin respect no-proxy-hosts when a proxy is used

### LINKED ISSUE #697: Gitlab plugin ignores NO_PROXY field of HTTP proxy settings
My Jenkins setup operates with set of intranet and internet resources. Internet access requires proxy settings. I added proxy configuration to Jenkins settings and also add all local resources to NO_PROXY list. With the configuration Gitlab plugin reports HTTP 401 error on each HTTP request to local Gitlab. Plugin works correctly, If I totally disable proxy.

The issue is localized at https://github.com/jenkinsci/gitlab-plugin/blob/316b8bebb8927081547d31333ec76e1c97b98f25/src/main/java/com/dabsquared/gitlabjenkins/gitlab/api/impl/ResteasyGitLabClientBuilder.java#L98

But it looks like RestEasyClientBuilder does not support NO_RPOXY. I'm ready to contribute fix, if you have any hint how to implement NO_PROXY logic.

### PR DESCRIPTION
**Pull request by Argelbargel**
Fixes #697 
This should fix #697 ensuring that the ResteasyClient-Builder respects non-proxy-hosts in setups using a proxy to connect to the internet

### CODE REVIEW

File: src/main/java/com/dabsquared/gitlabjenkins/gitlab/api/impl/ResteasyGitLabClientBuilder.java
Comment (Argelbargel) [inline src/main/java/com/dabsquared/gitlabjenkins/gitlab/api/impl/ResteasyGitLabClientBuilder.java]: This change ignores socks-proxy-setups as we currently force the proxy's protocol to http(s) anyway. Does anyone know whether the plugin ever worked with socks-proxys?

File: src/main/java/com/dabsquared/gitlabjenkins/gitlab/api/impl/ResteasyGitLabClientBuilder.java
Comment (Argelbargel) [inline src/main/java/com/dabsquared/gitlabjenkins/gitlab/api/impl/ResteasyGitLabClientBuilder.java:101]: This change ignores socks-proxy-setups as we currently force the proxy's protocol to http(s) anyway. Does anyone know whether the plugin ever worked with socks-proxys?

**Review by mreichel — APPROVED** [review]

File: src/main/java/com/dabsquared/gitlabjenkins/gitlab/api/impl/ResteasyGitLabClientBuilder.java
Comment (mreichel) [reply src/main/java/com/dabsquared/gitlabjenkins/gitlab/api/impl/ResteasyGitLabClientBuilder.java:101]: I don't know if it ever did.

Comment (omehegan) [reply src/main/java/com/dabsquared/gitlabjenkins/gitlab/api/impl/ResteasyGitLabClientBuilder.java:101]: Me either.

**Comment by omehegan** [comment]
I want to see if we get some more feedback in the original issue before I merge this.

File: src/main/java/com/dabsquared/gitlabjenkins/gitlab/api/impl/ResteasyGitLabClientBuilder.java
Comment (christ66) [inline src/main/java/com/dabsquared/gitlabjenkins/gitlab/api/impl/ResteasyGitLabClientBuilder.java:56]: Is it possible to instead rely on `java.net.Proxy.*` we could use the [Async Http Client Plugin](https://wiki.jenkins.io/display/JENKINS/Async+Http+Client+Plugin) (which IIRC has the proxy configuration in the UI under Manage Jenkins > Manage Plugins > Advanced tab)?

Comment (Argelbargel) [reply src/main/java/com/dabsquared/gitlabjenkins/gitlab/api/impl/ResteasyGitLabClientBuilder.java:56]: The UI unter Manage Plugins is the one built into Jenkins itself. ResteasyGitLabClientBuilder uses them already. What i could do is to use Proxy.Type.DIRECT instead of Proxy.Type.HTTP in the check to allow SOCKS-proxied connections but then we'd have to create a socks-configuration further down. As of relying on the Async Http Client Plugin - i'm not sure whether it is possible to use this with Resteasy

**Comment by omehegan** [comment]
Still waiting for feedback from the original issue submitter.

**Comment by omehegan** [comment]
This was just released in version 1.5.5 of the plugin. It should be available in the update center soon.

**Comment by naxim92** [comment]
Version 1.7.15 still no changes. Still ignoring NO_PROXY. See https://github.com/jenkinsci/gitlab-plugin/issues/1562

**Pull request merged**
